### ISSUE TITLE
Menu + Keyboard Shortcuts Overhaul

### ISSUE THREAD

**Issue by obuchtala** [substance/texture#866]
At the moment, there are two major things we should address here:
1. There is no consistent/complete toolbar, where the user can do everything that is possible without using shortcuts.
2. Keyboard shortcuts are defined all over the place, making this hard to maintain.

We should
1. provide an exhaustive solution using the toolbar and menus.
2. pull together all keyboard shortcut definitions and document them.

**Comment by michael** [comment]
Agree. Do you think we should have separate keymap files for each OS or put them into one and specify scopes. Many cases are addressed by using things like `CommandOrControl+S`.

We should do research on what shortcuts are used in Google Docs / Word etc.

**Comment by michael** [comment]
VS Code has a visualised key map, and they use scopes (editorFocus, — for global, textInputFocus etc) when a shortcut should trigger.

![image](https://user-images.githubusercontent.com/2931/45669287-dfec8480-bb1f-11e8-9dc6-3a223a2466ed.png)

I would start with organising ourselves better with regards to keymaps. We should introduce a help dialog that shows all the keyboard shortcuts. However I would not let user customise for now. Better have them make suggestions and we maintain the ultimate keymap.

**Comment by lpanebr** [comment]
Great. A save button will make it easier to debug.

Google Doc shortcuts list for different OSs is at https://support.google.com/docs/answer/179738?hl=en

Would it be better to put this info in a google sheet so we can add MS Office equivalents and people can comment...?

**Comment by michael** [comment]
@lpanebr yeah if you could set something up, that could serve as a base for discussion that would be extremely helpful! :)

**Comment by lpanebr** [comment]
@michael I've created the [Texture Shortcut Mappings sheet](https://docs.google.com/spreadsheets/d/1l4z_8El_2y_fXlqCX7ZJM5NQ9OE9PvRZieRzRnpd8hs/edit#gid=0) with Google Docs shortcut mappings for PC and Mac (Orange header columns). Unfortunately [MS Office shortcuts page](https://support.office.com/en-us/article/keyboard-shortcuts-for-microsoft-word-on-windows-95ef89dd-7142-4b50-afb2-f762f663ceb2#OfficeVersion=Newer_versions) is not table copy/paste friendly... 

The main columns are blue header:
Priority | Action | Keyboard Shortcut | Toolbar | Scopes | Comments

Maybe a net step would be to fill in what's already implemented in the Keyboard Shortcut and Toolbar columns and then figuring out how to collaborate from there?

**Comment by michael** [comment]
Thank you very much! That's a great foundation for discussion. Before we invite people to comment the next steps are:

1. Remove all actions that are not available in Texture (e.g. visual things like left-align, rotate etc)
2. Add Texture specific actions: Such as Add Reference, Add Blockquote etc.
3. Make an initial suggestion for shortcuts
4. Invite Consortium Members to comment

**Comment by michael** [comment]
@lpanebr btw the save button is now implemented.

**Comment by lpanebr** [comment]
I'll be happy to do 1, 2 and 3 but I need help to find out which actions are already available, since according to the OP they are "defined all over the place".

**Comment by oliver----** [comment]
I guess, what we need is a just a list of desired shortcuts. The decision what is already there is on us...

**Comment by lpanebr** [comment]
Okay. Yes, I thought this would be used to allow consortium members to suggest new actions and thus having the all inclusive list of actions (and shortcuts) available would help.

Anyway. I added the yellow column for the existing and desired Texture specific actions.

**So the intended use I imagine could be is:**
 1. the consortium members vote using the Priority column. Each member adds x's to actions. (If you want/need it's easy to setup something to allow voting per milestones..)
 2. the consortium members may add desired actions on the (new) yellow column
 3. the consortium members may make an initial suggestion for shortcuts in the comments

I also though the dev team could then use additional columns to define coding stuff like scopes etc and maybe use a more private version of a similar sheet to generate config files and even documentation.

Anyway fell free to make a copy and take over guys.
If I can help further just let me know. ;-)

**Comment by oliver----** [comment]
@lpanebr this is perfect!

**Comment by oliver----** [comment]
At some point we will convert this into something we can add to the repo, sacrificing editability.
Thank you!

**Comment by oliver----** [comment]
Do you mind to be the owner of the document?

**Comment by oliver----** [comment]
@michael can I give you the responsibility for this?

**Comment by michael** [comment]
Absolutely. Let's make this a priority for the Texture 2.0 release and take time to really consider user feedback.

**Comment by oliver----** [comment]
Yep. I can offer to add a mechanism to load settings from file.

**Comment by oliver----** [comment]
@lpanebr I need your full name because I'd like to add you to the list of Texture contributors.

**Comment by oliver----** [comment]
@lpanebr now you are an official contributor: https://github.com/substance/texture/commit/4920ad9c54fe50ad218ddb1ae69fd48d1a46ef61

😄

**Comment by lpanebr** [comment]
thank you! :smiley:
